onnxruntime/contrib_ops/cpu/bert/group_query_attention.cc

```diff
@@ -0,0 +1,198 @@
+  const bool packed_qkv = parameters.is_packed_qkv;
+
+  std::vector<int64_t> output_shape(3);
+  output_shape[0] = static_cast<int64_t>(batch_size);
+  output_shape[1] = static_cast<int64_t>(sequence_length);
+  output_shape[2] = static_cast<int64_t>(q_hidden_size);
+  Tensor* output = context->Output(0, output_shape);
+
+  std::vector<int64_t> present_k_shape({static_cast<int64_t>(batch_size), static_cast<int64_t>(kv_num_heads_), static_cast<int64_t>(present_kv_seqlen), static_cast<int64_t>(head_size)});
+  std::vector<int64_t> present_v_shape({static_cast<int64_t>(batch_size), static_cast<int64_t>(kv_num_heads_), static_cast<int64_t>(present_kv_seqlen), static_cast<int64_t>(head_size)});
+  Tensor* present_k = context->Output(1, present_k_shape);
+  Tensor* present_v = context->Output(2, present_v_shape);
+
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+
+  OrtValue Q;
+  OrtValue K;
+  OrtValue V;
+  if (packed_qkv) {
+    ORT_RETURN_IF_ERROR(MaybeTransposeToBNSH<T>(
+        allocator, batch_size, num_heads_ + 2 * kv_num_heads_, sequence_length, head_size, query, Q));
+  } else {
+    ORT_RETURN_IF_ERROR(MaybeTransposeToBNSH<T>(
+        allocator, batch_size, num_heads_, sequence_length, head_size, query, Q));
+    ORT_RETURN_IF_ERROR(MaybeTransposeToBNSH<T>(
+        allocator, batch_size, kv_num_heads_, sequence_length, head_size, key, K));
+    ORT_RETURN_IF_ERROR(MaybeTransposeToBNSH<T>(
+        allocator, batch_size, kv_num_heads_, sequence_length, head_size, value, V));
+  }
+
+  if (do_rotary_) {
+    rotary_embedding_helper::RotaryParameters rotary_params = {};
+    rotary_params.batch_size = batch_size;
+    rotary_params.sequence_length = sequence_length;
+    rotary_params.hidden_size = q_hidden_size;
+    rotary_params.head_size = head_size;
+    rotary_params.rotary_embedding_dim = parameters.rotary_dim;
+    rotary_params.num_heads = num_heads_;
+    rotary_params.max_sequence_length = sequence_length;  // unused
+    rotary_params.seq_stride = head_size;
+    rotary_params.head_stride = sequence_length * rotary_params.seq_stride;
+    rotary_params.batch_stride = (packed_qkv ? (num_heads_ + 2 * kv_num_heads_) : num_heads_) * rotary_params.head_stride;
+    rotary_params.position_ids_format = 1;
+    rotary_params.transposed = true;
+    auto* tp = context->GetOperatorThreadPool();
+    OrtValue pos_ids;
+    auto element_type = DataTypeImpl::GetType<int64_t>();
+    std::vector<int64_t> pos_id_dim({batch_size, sequence_length});
+    gsl::span<const int64_t> pos_id_dim_span{pos_id_dim};
+    TensorShape pos_id_shape(pos_id_dim_span);
+    Tensor::InitOrtValue(element_type, pos_id_shape, allocator, pos_ids);
+    ORT_RETURN_IF_ERROR(group_query_attention_helper::GeneratePositionIds(seqlens_k->Data<int32_t>(), batch_size,
+                                                                          sequence_length, pos_ids.GetMutable<Tensor>()->MutableData<int64_t>(), context));
```

Comment:
why do we need to generate a position id?

Reply:
supposedly, it is a increasing sequence

Reply:
actually we can not generate in the case of prompt, but in case of token gen we need to use a int64_t cast of seqlens_k since the new tokens need to be correctly index into the sequence.